### PR TITLE
Fix out= function semantics for named tensors.

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24306 Implement name inference for mm, addmm
* #24203 Implement name inference for t(), transpose(...)
* #24087 Name inference for softmax, log_softmax and Dimname overloads.
* **#24028 Fix out= function semantics for named tensors.**
* #23973 Update tensor.view_names / tensor.names_ API
* #23962 Rename set_names -> view_names, set_names_ -> names_
* #23804 Implement tensor.align_to(names), torch.align_tensors(t1, t2, ...)
* #24109 Named inference for contiguous(), bernoulli variants, and dropout.
* #24108 Implement name inference rule for empty_like, clone
* #24107 Add `names` argument to ones, rand, randn, zeros, full; fix empty

Previously, torch.abs(tensor, out=out) would ignore the names of the
`out` tensor and overwrite them with the names of `tensor`.

This patch changes the behavior to the following:
1) If `out` does not have names, then overwite them with `tensor.names`.
2) If `out` does have names, then check that `out.names` equals
`tensor.names`.

This patch also includes the following clean ups:
- renamed `default_names` to `FIXME_default_names` because it is
inefficient and needs to be fixed.
- Renamed impl::internal_get_names / impl::internal_has_names to
impl::get_names / impl::set_names. Devs should feel free to use them, so
I removed the internal_ prefix.
- Moved internal_set_names to NamedTensor.{h, cpp}. These functions
still have the internal_ prefix because their use requires caution.

Test Plan:
- [namedtensor ci]

Differential Revision: [D16763387](https://our.internmc.facebook.com/intern/diff/D16763387)

### CODE REVIEW

File: aten/src/ATen/NamedTensorUtils.cpp [hunk 1]
```diff
@@ -174,6 +160,42 @@ static std::bitset<64> compute_included_idxs(IntArrayRef excluded_idxs) {
   return included_idxs;
 }
 
+static void assert_names_equal(DimnameList a, DimnameList b) {
+  TORCH_CHECK(a == b,
```
Comment (nairbv) [inline aten/src/ATen/NamedTensorUtils.cpp]: What does this `==` do if the `out` tensor is named with wildcard names?

Comment (nairbv) [reply aten/src/ATen/NamedTensorUtils.cpp]: or does that not matter because you propagate before you call this assertion?

Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.cpp]: the `==` checks that the names are literally the same. It does not do a "matching check".

So, for example, the following are disallowed:
- `torch.abs(Tensor[N, C], out=Tensor[N, None])`
- `torch.abs(Tensor[N, None], out=Tensor[N, C])`

Comment (nairbv) [reply aten/src/ATen/NamedTensorUtils.cpp]: Do we plan to add support for more combinations later or does outputting to a partially named tensor not make sense?

Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.cpp]: > Do we plan to add support for more combinations later or  does outputting to a partially named tensor not make sense?

It's possible to modify how all of this works by changing this line of the code. I am open to changing the semantics if it makes sense, but right now I am not sure if it makes sense so I put in a pretty conservative check.

**Review by nairbv — APPROVED** [review]

File: aten/src/ATen/NamedTensorUtils.cpp (hunk 1 above)
Comment (nairbv) [reply aten/src/ATen/NamedTensorUtils.cpp]: makes sense. I guess I was expecting it to follow the same kind of matching rules as other ops, but can always be relaxed later.

Comment (zou3519) [reply aten/src/ATen/NamedTensorUtils.cpp]: Our semantics for output tensors are a little weird in general. `out=` tensors get automatically resized to what they need to be.

One concrete example is that you can't resize output tensors that are also input tensors:
```
import torch
x = torch.randn(3, 3)
y = torch.randn(3)
y += x
```
If you can't change the size of an input tensor that is also an output tensor, we probably shouldn't allow the user do this for names. The current conservative check enforces this, but we could relax it a little and still have it enforce this.
